core/src/unprocessed_packet_batches.rs

```diff
@@ -325,27 +330,64 @@ pub fn transactions_to_deserialized_packets(
 // This function deserializes packets into transactions, computes the blake3 hash of transaction
 // messages, and verifies secp256k1 instructions. A list of sanitized transactions are returned
 // with their packet indexes.
-#[allow(clippy::needless_collect)]
 pub fn transaction_from_deserialized_packet(
     deserialized_packet: &ImmutableDeserializedPacket,
     feature_set: &Arc<feature_set::FeatureSet>,
     votes_only: bool,
     address_loader: impl AddressLoader,
+    tx_account_lock_limit: usize,
+    transaction_error_counters: &mut TransactionErrorMetrics,
 ) -> Option<SanitizedTransaction> {
     if votes_only && !deserialized_packet.is_simple_vote() {
         return None;
     }
-    let tx = SanitizedTransaction::try_new(
-        deserialized_packet.transaction().clone(),
-        *deserialized_packet.message_hash(),
-        deserialized_packet.is_simple_vote(),
+
+    // Check that the account locks are valid
+    let message = SanitizedMessage::try_new(
+        deserialized_packet.transaction().get_message().clone(),
         address_loader,
     )
     .ok()?;
+    match SanitizedTransaction::validate_account_locks(&message, tx_account_lock_limit) {
+        Err(TransactionError::AccountLoadedTwice) => {
+            saturating_add_assign!(transaction_error_counters.total, 1);
+            saturating_add_assign!(transaction_error_counters.account_loaded_twice, 1);
+        }
+        Err(TransactionError::TooManyAccountLocks) => {
+            saturating_add_assign!(transaction_error_counters.total, 1);
+            saturating_add_assign!(transaction_error_counters.too_many_account_locks, 1);
+        }
+        _ => {}
+    }
+
+    let tx = SanitizedTransaction::new(
+        deserialized_packet.transaction().clone(),
+        message,
+        *deserialized_packet.message_hash(),
+        deserialized_packet.is_simple_vote(),
+    );
+
     tx.verify_precompiles(feature_set).ok()?;
     Some(tx)
 }
 
+pub fn transaction_from_deserialized_packet_without_metrics(
+    deserialized_packet: &ImmutableDeserializedPacket,
+    feature_set: &Arc<feature_set::FeatureSet>,
+    votes_only: bool,
+    address_loader: impl AddressLoader,
+    tx_account_lock_limit: usize,
+) -> Option<SanitizedTransaction> {
```

Comment:
is this function just for test? If so, can rename it to `transaction_from_deserialized_packet_for_test` for clarify. Or wdyt of making `TransactionErrorMetrics` optional?